syntax = "proto3";
package fx.dex.v1;

import "gogoproto/gogo.proto";
import "fx/dex/v1/order.proto";
import "cosmos/base/v1beta1/coin.proto";
import "cosmos/msg/v1/msg.proto";

option go_package = "github.com/marginxio/marginx/x/dex/v1/types";

service Msg {
  rpc CreateOrder(MsgCreateOrder) returns (MsgCreateOrderResponse);
  rpc CancelOrder(MsgCancelOrder) returns (MsgCancelOrderResponse);

  rpc AddMargin(MsgAddMargin) returns (MsgAddMarginResp);
  rpc ClosePosition(MsgClosePosition) returns (MsgClosePositionResp);
  rpc LiquidationPosition(MsgLiquidationPosition)
      returns (MsgLiquidationPositionResp);
  rpc FundDexPool(MsgFundDexPool) returns (MsgFundDexPoolResp);
}

message MsgCreateOrder {
  option (cosmos.msg.v1.signer) = "owner";

  string owner = 1;
  string pair_id = 2;
  Direction direction = 3;
  string price = 4 [
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable) = false
  ];
  string base_quantity = 5 [
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable) = false
  ];
  int64 leverage = 6;
}

message MsgCreateOrderResponse { string order_id = 1; }

message MsgCancelOrder {
  option (cosmos.msg.v1.signer) = "owner";

  string owner = 1;
  string order_id = 2;
}

message MsgCancelOrderResponse {}

message MsgAddMargin {
  option (cosmos.msg.v1.signer) = "owner";

  string owner = 1;
  string pair_id = 2;
  string position_id = 3 [
    (gogoproto.customtype) = "cosmossdk.io/math.Uint",
    (gogoproto.nullable) = false
  ];
  string margin = 4 [
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable) = false
  ];
}

message MsgAddMarginResp {}

message MsgClosePosition {
  option (cosmos.msg.v1.signer) = "owner";

  string owner = 1;
  string pair_id = 2;
  string position_id = 3 [
    (gogoproto.customtype) = "cosmossdk.io/math.Uint",
    (gogoproto.nullable) = false
  ];
  string price = 4 [
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable) = false
  ];
  string base_quantity = 5 [
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable) = false
  ];
  bool full_close = 6;
  bool market_close = 7;
}

message MsgClosePositionResp { string order_id = 1; }

message MsgLiquidationPosition {
  option (cosmos.msg.v1.signer) = "liquidator";

  string liquidator = 1;
  repeated string position_ids = 2 [
    (gogoproto.customtype) = "cosmossdk.io/math.Uint",
    (gogoproto.nullable) = false
  ];
}

message MsgLiquidationPositionResp {}

message MsgFundDexPool {
  option (cosmos.msg.v1.signer) = "depositor";

  repeated cosmos.base.v1beta1.Coin amount = 1 [
    (gogoproto.nullable) = false,
    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
  ];
  string depositor = 2;
}

message MsgFundDexPoolResp {}
